factor/sms/classes/factor.php

```diff
@@ -0,0 +1,241 @@
+        // It must only grab the record with the user agent as the label.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+               AND label IS NOT NULL';
+
+        $record = $DB->get_record_sql($sql, array($USER->id, 'sms'));
+        $duration = get_config('factor_sms', 'duration');
+        $newcode = random_int(100000, 999999);
+
+        $number = $USER->phone2;
+
+        if (empty($record)) {
+            // No code active, generate new code.
+            $instanceid = $DB->insert_record('tool_mfa', array(
+                'userid' => $USER->id,
+                'factor' => 'sms',
+                'secret' => $newcode,
+                'label' => $number,
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ), true);
+            $this->sms_verification_code($instanceid);
+
+        } else if ($record->timecreated + $duration < time()) {
+            // Old code found. Keep id, update fields.
+            $DB->update_record('tool_mfa', array(
+                'id' => $record->id,
+                'secret' => $newcode,
+                'label' => $number,
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ));
+            $instanceid = $record->id;
+            $this->sms_verification_code($instanceid);
+        }
+    }
+
+    private function sms_verification_code($instanceid) {
+        global $DB;
+
+        // Here we should get the information, then construct the message.
+        $instance = $DB->get_record('tool_mfa', ['id' => $instanceid]);
+        $user = \core_user::get_user($instance->userid);
+        $phonenumber = $user->phone2;
+        $gateway = new \factor_sms\local\smsgateway\aws_sns();
+        $gateway->send_sms_message($instance->secret, $phonenumber);
+    }
+
+    /**
+     * Verifies entered code against stored DB record.
+     *
+     * @return bool
+     */
+    private function check_verification_code($enteredcode) {
+        global $DB, $USER;
+        $duration = get_config('factor_email', 'duration');
+
+        // Get instance that isnt parent email type (label check).
+        // This check must exclude the main singleton record, with the label as the email.
+        // It must only grab the record with the user agent as the label.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+               AND label IS NOT NULL';
+        $record = $DB->get_record_sql($sql, array($USER->id, 'sms'));
+
+        if ($enteredcode == $record->secret) {
```

Comment:
I'm not sure we should store the code here. Scenario is a little far fetched but real:

1) phone is stolen by mallory, notifications are shown on the lock screen (bad but not uncommon)
2) alice gets a phishing attack from mallory
3) she logs in to fake site (bad) so now mallory has the password. Up until now its non mfa so mfa should mitigate this
4) mallory gets the sms code and they are in because it works for their session. MFA has not helped

Unlike TOTP we can tie the sms code to the session instead of storing it so now the code is useless to mallory. And we can also go a step further and treat a code use by a different session as suspicious or an explicit fail rather than a neutral

Reply:
except no in this case mallory could replay both the password and also now get the sms text directly so it won't really help. hhmmm maybe overkill